### ISSUE TITLE
Type aliases are not recognized correctly in some situations with import cycles

### ISSUE THREAD

**Issue by dfroger** [python/mypy#5130]
This may be a duplicate issue, but I'm a bit lost... I'm using `mypy 0.600` and `Python 3.6.2`.

```
# a.py
from typing import Union

from c import T

class A:
    x: str

class B:
    x: str

U = Union[A, B]
```

```
# b.py
from c import U

def foo(u: U):
    print(u.x)
```

```
# c.py
class T:
    pass

from a import U
from b import foo
```

`python -c 'import c'` works, `Mypy .` give the errors:

```
b.py:3: error: Invalid type "a.U"
b.py:4: error: U? has no attribute "x"
```

I would like to understand how here `Mypy` import mechanism is different from `Python` import mechanism, and why `Mypy` fails with `U` type.

**Comment by JelleZijlstra** [comment]
Mypy doesn't exactly replicate Python's mechanism for resolving circular imports. Python can basically move back and forth between modules during execution; mypy doesn't do that.

This seems closely similar to #4930, although that one was triggered only by `import *`.

**Comment by dfroger** [comment]
Also maybe similar to #4442

**Comment by andy-ms** [comment]
Ran into a similar issue:

**a.py**
```py
def incr(i: int) -> int:
    from b import Integer
    res: Integer = i + 1
    return res
```

**b.py**
```py
from a import incr

Integer = int
```

Results in an error: `a.py:4: error: Invalid type "Integer"`

**Comment by ilevkivskyi** [comment]
I can't reproduce the original example on master. The new example still fails, but passes with the `--new-semantic-analyzer` flag. This should be closed when we release the version with new analyzer.

**Comment by ilevkivskyi** [comment]
Closing as this is fixed with new semantic analyzer (will be released soon).
